streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java

```diff
@@ -62,206 +64,229 @@
-        collector.send("topic1", "9", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
-        collector.send("topic1", "27", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
-        collector.send("topic1", "81", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
-        collector.send("topic1", "243", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+        assertEquals(4L, (long) offsets.get(new TopicPartition(topic, 0)));
+        assertEquals(2L, (long) offsets.get(new TopicPartition(topic, 1)));
+        assertEquals(0L, (long) offsets.get(new TopicPartition(topic, 2)));
+        assertEquals(9, producer.history().size());
 
-        collector.send("topic1", "28", "0", headers, null, stringSerializer, stringSerializer, streamPartitioner);
-        collector.send("topic1", "82", "0", headers, null, stringSerializer, stringSerializer, streamPartitioner);
-        collector.send("topic1", "244", "0", headers, null, stringSerializer, stringSerializer, streamPartitioner);
+        // returned offsets should not be modified
+        final TopicPartition topicPartition = new TopicPartition(topic, 0);
+        assertThrows(UnsupportedOperationException.class, () -> offsets.put(topicPartition, 50L));
+    }
 
-        collector.send("topic1", "245", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+    @Test
+    public void shouldSendWithNoPartition() {
+        final Headers headers = new RecordHeaders(new Header[]{new RecordHeader("key", "value".getBytes())});
+
+        collector.send(topic, "3", "0", headers, null, null, stringSerializer, stringSerializer);
+        collector.send(topic, "9", "0", headers, null, null, stringSerializer, stringSerializer);
+        collector.send(topic, "27", "0", headers, null, null, stringSerializer, stringSerializer);
+        collector.send(topic, "81", "0", headers, null, null, stringSerializer, stringSerializer);
+        collector.send(topic, "243", "0", headers, null, null, stringSerializer, stringSerializer);
+        collector.send(topic, "28", "0", headers, null, null, stringSerializer, stringSerializer);
+        collector.send(topic, "82", "0", headers, null, null, stringSerializer, stringSerializer);
+        collector.send(topic, "244", "0", headers, null, null, stringSerializer, stringSerializer);
+        collector.send(topic, "245", "0", headers, null, null, stringSerializer, stringSerializer);
 
         final Map<TopicPartition, Long> offsets = collector.offsets();
 
-        assertEquals((Long) 4L, offsets.get(new TopicPartition("topic1", 0)));
-        assertEquals((Long) 2L, offsets.get(new TopicPartition("topic1", 1)));
-        assertEquals((Long) 0L, offsets.get(new TopicPartition("topic1", 2)));
+        // with mock producer without specific partition, we would use default producer partitioner with murmur hash
```

Comment:
@mjsax this test is added as suggested by @abbccdda , and it actually covers the case that we may want to fix later. So it is also used as a "guard" which would fail when we do not use default partitioner of producer.